internal/k8s/reconciler/listener.go

```diff
@@ -119,12 +140,86 @@ func (l *K8sListener) validateTLS(ctx context.Context) error {
 		}
 		l.status.ResolvedRefs.InvalidCertificateRef = certificateErr
 	} else {
-		l.certificates = []string{resource}
+		if l.tlsParams != nil {
+			l.tlsParams.certificates = []string{resource}
+		} else {
+			l.tlsParams = &tlsParams{
+				certificates: []string{resource},
+			}
+		}
+	}
+
+	if l.listener.TLS.Options != nil {
+		tlsMinVersion := l.listener.TLS.Options[tlsMinVersionAnnotationKey]
+		tlsMaxVersion := l.listener.TLS.Options[tlsMaxVersionAnnotationKey]
+		tlsCipherSuitesStr := l.listener.TLS.Options[tlsCipherSuitesAnnotationKey]
+
+		if tlsMinVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMinVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS min version")
+				return nil
+			}
+
+			if tlsCipherSuitesStr != "" {
+				if _, ok := tlsVersionsWithConfigurableCipherSuites[string(tlsMinVersion)]; !ok {
+					l.status.Ready.Invalid = errors.New("configuring TLS cipher suites is only supported for TLS 1.2 and earlier")
+					return nil
+				}
+			}
+		}
+
+		if tlsMaxVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMaxVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS max version")
+				return nil
+			}
+		}
+
+		if tlsCipherSuitesStr != "" {
+			// split comma delimited string into string array and trim whitespace
+			tlsCipherSuitesUntrimmed := strings.Split(string(tlsCipherSuitesStr), ",")
+			tlsCipherSuites := tlsCipherSuitesUntrimmed[:0]
+			for _, c := range tlsCipherSuitesUntrimmed {
+				tlsCipherSuites = append(tlsCipherSuites, strings.TrimSpace(c))
+			}
+
+			// validate each cipher suite in array
+			for _, c := range tlsCipherSuites {
+				if _, ok := supportedTlsCipherSuites[c]; !ok {
+					l.status.Ready.Invalid = fmt.Errorf("unrecognized or unsupported TLS cipher suite: %s", c)
+					return nil
+				}
+
+				// TODO: append to Listener TLSParams.CipherSuites?
+			}
+		}
 	}
 
 	return nil
 }
 
+// FIXME: import this from Consul API
+var supportedTlsVersions = map[string]struct{}{
+	"TLS_AUTO": {},
+	"TLSv1_0":  {},
+	"TLSv1_1":  {},
+	"TLSv1_2":  {},
+	"TLSv1_3":  {},
+}
+
+// FIXME: import this from Consul API
+var tlsVersionsWithConfigurableCipherSuites = map[string]struct{}{
+	"TLS_AUTO": {}, // Remove if Envoy ever sets TLS 1.3 as default minimum
+	"TLSv1_0":  {},
+	"TLSv1_1":  {},
+	"TLSv1_2":  {},
+}
+
+// FIXME: import this from Consul API
+var supportedTlsCipherSuites = map[string]struct{}{
+	"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": {},
+}
```

Comment:
Would it be appropriate to import these lookup tables from the Consul API? I'd hate the have to maintain them independently, although having them in some shared Envoy/xDS config dependency will probably make more sense in the future than accessing through Consul core API.

Reply:
I suppose it all depends on how these are exposed in Consul. I like the idea of importing from there to keep things a bit cleaner, at least so long as we don't diverge from what the core codebase supports -- but are these exposed in the Consul subpackage?

Reply:
I really wanted to expose these, but the Consul codebase doesn't allow sharing anything between `consul` and `consul/api` (even things like types or struct defs) from either package to the other, which seems to make this not possible. In the future it would be nice to have this list as part of a common xDS library used by both Consul core and Consul API Gateway.